pkg/apis/extensions/v1alpha1/types_dnsrecord.go

```diff
@@ -0,0 +1,111 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ Object = (*DNSRecord)(nil)
+
+// DNSRecordResource is a constant for the name of the DNSRecord resource.
+const DNSRecordResource = "DNSRecord"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced,path=DNSRecords,shortName=cp,singular=DNSRecord
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name=Type,JSONPath=".spec.type",type=string,description="The control plane type."
+// +kubebuilder:printcolumn:name=Purpose,JSONPath=".spec.purpose",type=string,description="Purpose of control plane resource."
+// +kubebuilder:printcolumn:name=Status,JSONPath=".status.lastOperation.state",type=string,description="Status of control plane resource."
+// +kubebuilder:printcolumn:name=Age,JSONPath=".metadata.creationTimestamp",type=date,description="creation timestamp"
+
+// DNSRecord is a specification for a DNSRecord resource.
+type DNSRecord struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec DNSRecordSpec `json:"spec"`
+	// +optional
+	Status DNSRecordStatus `json:"status"`
+}
+
+// GetExtensionSpec implements Object.
+func (i *DNSRecord) GetExtensionSpec() Spec {
+	return &i.Spec
+}
+
+// GetExtensionStatus implements Object.
+func (i *DNSRecord) GetExtensionStatus() Status {
+	return &i.Status
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DNSRecordList is a list of DNSRecord resources.
+type DNSRecordList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is the list of DNSRecords.
+	Items []DNSRecord `json:"items"`
+}
+
+// DNSRecordSpec is the spec of a DNSRecord resource.
+type DNSRecordSpec struct {
+	// DefaultSpec is a structure containing common fields used by all extension resources.
+	DefaultSpec `json:",inline"`
+	// SecretRef is a reference to a secret that contains the cloud provider specific credentials.
+	SecretRef corev1.SecretReference `json:"secretRef"`
+	// Region is the region of this DNS record. If not specified, the region specified in SecretRef will be used.
+	// If that is also not specified, a certain default region will be used, e.g. us-west-2 for AWS route53.
```

Comment:
Nit, we should probably not include implementation specifics that quickly become outdated in the generic docs.
E.g. like this?
```suggestion
	// If that is also not specified, the extension controller will use its default region.
```

Reply:
Done